fhirpipe/transform/dataframe.py

```diff
@@ -0,0 +1,73 @@
+import numpy as np
+
+from fhirpipe.utils import new_col_name
+from fhirpipe.scripts import get_script
+
+
+def squash_rows(df, squash_rules, parent_cols=[], squash_aff_func=None):
+    """
+    Apply the OneToMany joins to have a single result with a list in it from
+    a list of "flat" results.
+
+    args:
+        rows (list<str>): all the results returned from a sql query
+        squash_rules (tuple<list>): which columns should serve as identifier to
+        merge the rows
+        parent_cols (list): param used for recursive call
+
+    Example:
+        if you join people with bank accounts on guy.id = account.owner_id,
+        you want at the end to have for a single guy to have a single instance
+        with an attribute accounts.
+        ROWS:
+        GUY.NAME    ...     GUY.AGE     ACCOUNT.NAME        ACCOUNT.AMOUNT
+        Robert              21          Compte courant      17654
+        Robert              21          Compte d'epargne    123456789
+        David               51          Ibiza summer        100
+
+        Squash rule:
+        ('GUY', 'ACCOUNT') or in terms of columns ([0, ..., 5], [6, 7])
+
+        Output:
+        GUY.NAME    ...     GUY.AGE     ACCOUNT.NAME        ACCOUNT.AMOUNT
+        Robert              21        [(Compte courant   ,  17654         )
+                                       (Compte d'epargne ,  123456789     )]
+        David               51          Ibiza summer        100
+    """
+    table, child_rules = squash_rules
+
+    new_cols = [col for col in df.columns if col.startswith(f"{table}.")]
+    pivot_cols = parent_cols + new_cols
+
+    to_squash = [col for col in df.columns if col not in pivot_cols]
+
+    if not to_squash:
+        return df
+
+    for child_rule in child_rules:
+        df = squash_rows(df, child_rule, pivot_cols)
+
+    if squash_aff_func is None:
+        squash_aff_func = squash_agg_first
+
+    df = df.groupby(pivot_cols, as_index=False)[to_squash].agg(squash_aff_func)
```

Comment:
Why not directly do `.add(squash_agg_first)` ? (and maybe say in 1 line what it does)

Reply:
It was to be able to pass another aggregation function.
But I'm still not sure how the user could know how to squash rows...

Reply:
Yes maybe that's asking a lot to the user, and any way the elements in the pivot cols should be the same (it's a left join) and taking any row should be equivalent

Reply:
Yes, but here, the aggregation function is applied to the non-pivot columns.
For instance, an error has been made when filling the database and we have:
PIVOT | NATIONALITY
Bob     | French
Bob     | American
we don't know how to aggregate French and American

I want to default to taking the first non-null value but I'm not sure we want to silently pass on ambiguous cases